### PR TITLE
Inheritance structure for multioutput code + updated multioutput notebook

### PR DESCRIPTION
**Pull request by markvdw**
Tests still need work, but the notebook works.

### CODE REVIEW

File: gpflow/kernels/mo_kernels.py [hunk 1]
```diff
@@ -130,7 +130,18 @@ def K_diag(self, X, full_output_cov=False, presliced=False):
             stacked) if full_output_cov else stacked  # [N, P, P]  or  [N, P]
 
 
-class SeparateMixedMok(Mok, Combination):
+class IndependentLatentBase(Mok, metaclass=abc.ABCMeta):
```
Comment (st--) [inline gpflow/kernels/mo_kernels.py]: If we have `Mok` in the class name everywhere else would do that here for consistency, too

Comment (markvdw) [reply gpflow/kernels/mo_kernels.py]: I'm sorting out naming in the next PR, on Artem's request.

File: _unsorted/_test_methods.py [hunk 2]
```diff
@@ -63,13 +63,13 @@ def _check_models_close(m1, m2, tolerance=1e-2):
     gpflow.models.VGP(Datum.X, Datum.Y, Datum.kernel, Datum.lik),
     gpflow.models.GPMC(Datum.X, Datum.Y, Datum.kernel, Datum.lik),
     gpflow.models.SGPMC(Datum.X, Datum.Y, Datum.kernel, Datum.lik, features=Datum.Z),
```
Comment (st--) [inline _unsorted/_test_methods.py]: should this be consistently renamed as well ?

Comment (st--) [reply _unsorted/_test_methods.py]: (grep for `feature`?)

File: gpflow/covariances/mo_kuus.py [hunk 3]
```diff
@@ -2,71 +2,70 @@
...
-@Kuu.register(MixedKernelSharedMof, SeparateMixedMok)
-def _Kuu(feature: MixedKernelSharedMof,
-         kernel: SeparateMixedMok,
-         *,
-         jitter=0.0):
-    Kmm = tf.stack([Kuu(feature.feature, k) for k in kernel.kernels],
-                   axis=0)  # [L, M, M]
-    jittermat = tf.eye(len(feature), dtype=Kmm.dtype)[None, :, :] * jitter
-    return Kmm + jittermat
+# @Kuu.register(MixedKernelSharedMof, SeparateMixedMok)
```
Comment (st--) [inline gpflow/covariances/mo_kuus.py]: Remove commented-out code

File: gpflow/covariances/mo_kuus.py [hunk 4]
```diff
@@ -2,71 +2,70 @@
...
                    axis=0)  # [L, M, M]
     jittermat = tf.eye(len(feature), dtype=Kmm.dtype)[None, :, :] * jitter
     return Kmm + jittermat
 
 
-@Kuu.register(SeparateIndependentMof, SharedIndependentMok)
-def _Kuu(feature: SeparateIndependentMof,
-         kernel: SharedIndependentMok,
+@Kuu.register(FallbackSeparateIndependentInducingVariables, SharedIndependent)
+def _Kuu(feature: FallbackSeparateIndependentInducingVariables,
```
Comment (st--) [inline gpflow/covariances/mo_kuus.py]: feature -> inducing_variable for consistency (possibly elsewhere, too)

File: gpflow/covariances/mo_kuus.py [hunk 5]
```diff
@@ -2,71 +2,70 @@
...
+         kernel: SharedIndependent,
          *,
          jitter=0.0):
-    Kmm = tf.stack([Kuu(f, kernel.kernel) for f in feature.features],
+    Kmm = tf.stack([Kuu(f, kernel.kernel) for f in feature.inducing_variable_list],
```
Comment (st--) [inline gpflow/covariances/mo_kuus.py]: Should `kernel.kernel` (and `kernel.kernels`) be tidied up likewise ?

File: gpflow/expectations/products.py [hunk 6]
```diff
@@ -59,7 +59,7 @@ def _E(p, kern1, feat1, kern2, feat2, nghp=None):
     :return: NxMxM
     """
     if feat1 != feat2:
-        raise NotImplementedError("Different features are not supported.")
+        raise NotImplementedError("Different inducing are not supported.")
```
Comment (st--) [inline gpflow/expectations/products.py]: inducing *variables*?

Comment (st--) [reply gpflow/expectations/products.py]: also elsewhere

File: gpflow/inducing_variables/inducing_variables.py [hunk 7]
```diff
@@ -20,21 +20,21 @@
 from ..config import default_float
 
 
-class InducingFeature(tf.Module):
+class InducingVariables(tf.Module):
```
Comment (st--) [inline gpflow/inducing_variables/inducing_variables.py]:
```suggestion
class InducingVariable(tf.Module):
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    `Kuu()` and `Kuf()` return. This allows a custom `conditional()` to provide
+    the most efficient implementation.
+    """
+    pass
+
+
+class SeparateIndependentInducingVariables(FallbackSeparateIndependentInducingVariables):
+    """
+    Here, we define the same inducing variables as in the base class. However,
+    this class is intended to be used without the constrains on the shapes that
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
    this class is intended to be used without the constraints on the shapes that
```

File: tests/test_multioutput.py [hunk 9]
```diff
@@ -450,7 +450,7 @@ def test_separate_independent_mof():
     q_mu_1 = np.random.randn(Data.M * Data.P, 1)
     q_sqrt_1 = np.tril(np.random.randn(Data.M * Data.P, Data.M * Data.P))[None, ...]  # 1 x MP x MP
 
-    kernel_1 = mk.SharedIndependentMok(RBF(variance=0.5, lengthscale=1.2), Data.P)
+    kernel_1 = mk.SharedIndependent(RBF(variance=0.5, lengthscale=1.2), Data.P)
```
Comment (st--) [inline tests/test_multioutput.py]:
```suggestion
    kernel_1 = mk.SharedIndependent(SquaredExponential(variance=0.5, lengthscale=1.2), Data.P)
```

File: tests/test_multioutput.py [hunk 10]
```diff
@@ -466,9 +466,9 @@ def closure1():
     # Model 2 (efficient)
     q_mu_2 = np.random.randn(Data.M, Data.P)
     q_sqrt_2 = np.array([np.tril(np.random.randn(Data.M, Data.M)) for _ in range(Data.P)])  # P x M x M
-    kernel_2 = mk.SharedIndependentMok(RBF(variance=0.5, lengthscale=1.2), Data.P)
+    kernel_2 = mk.SharedIndependent(RBF(variance=0.5, lengthscale=1.2), Data.P)
     feat_list_2 = [InducingPoints(Data.X[:Data.M, ...]) for _ in range(Data.P)]
```
Comment (st--) [inline tests/test_multioutput.py]:
```suggestion
    feat_list_2 = [InducingPoints(Data.X[:Data.M, ...]) for _ in range(Data.P)]
```
```suggestion
    inducing_variable_list_2 = [InducingPoints(Data.X[:Data.M, ...]) for _ in range(Data.P)]
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+class MultioutputInducingVariables(InducingVariables):
+    """
+    Multioutput Inducing Variables
+    Base class for methods which define a collection of inducing variables which
+    in some way can be grouped. The main example is where the inducing variables
+    consist of outputs of various independent GPs. This can be because our model
+    uses multiple independent GPs (SharedIndependentMok, SeparateIndependentMok)
+    or because it is constructed from independent GPs (eg IndependentLatentBase,
+    SeparateMixedMok).
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]: LinearCoregionalisation kernel ?

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+from .inducing_variables import InducingVariables
+
+
+class MultioutputInducingVariables(InducingVariables):
+    """
+    Multioutput Inducing Variables
+    Base class for methods which define a collection of inducing variables which
+    in some way can be grouped. The main example is where the inducing variables
+    consist of outputs of various independent GPs. This can be because our model
+    uses multiple independent GPs (SharedIndependentMok, SeparateIndependentMok)
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]: update

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    or because it is constructed from independent GPs (eg IndependentLatentBase,
+    SeparateMixedMok).
+    """
+    pass
+
+
+class FallbackSharedIndependentInducingVariables(MultioutputInducingVariables):
+    """
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]: Should have the specific thing ("what is 'shared'") up front, not hidden at end of docstring.

Also, lots of redundancy with the Separate... class: can this be unified ?

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    from independent latent processes acutally allow even more efficient
+    approximations. However, we include this code path, as it does not require
+    specifying a new `conditional()` implementation.
+
+    Here, we share the definition of inducing variables between all latent
+    processes.
+    """
+
+    def __init__(self, inducing_variable):
+        MultioutputInducingVariables.__init__(self)
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
        super().__init__()
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        self.inducing_variable = inducing_variable
```
Comment (st--) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
        self.shared_inducing_variable = inducing_variable
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    or because it is constructed from independent GPs (eg IndependentLatentBase,
+    SeparateMixedMok).
+    """
+    pass
+
+
+class FallbackSharedIndependentInducingVariables(MultioutputInducingVariables):
+    """
+    This class is designated to be used to
+     - provide a general a general interface for multioutput kernels
```
Comment (vdutor) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
     - provide a general interface for multioutput kernels
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+       constructed from independent latent processes,
+     - only require the specification of Kuu and Kuf.
+    All multioutput kernels constructed from independent latent processes allow
+    the inducing variables to be specified in the latent processes, and a
+    reasonably efficient method (i.e. one that takes advantage of the
+    independence in the latent processes) can be specified quite generally by
+    only requiring the following covariances:
+     - Kuu: [L1, M, M]    (L can be 1, the conditional can broadcast),
+     - Kuf: [L, M, N, P].
```
Comment (vdutor) [inline gpflow/inducing_variables/mo_inducing_variables.py]: Can `L` be `1` here is well?

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    This class is designated to be used to
+     - provide a general a general interface for multioutput kernels
+       constructed from independent latent processes,
+     - only require the specification of Kuu and Kuf.
+    All multioutput kernels constructed from independent latent processes allow
+    the inducing variables to be specified in the latent processes, and a
+    reasonably efficient method (i.e. one that takes advantage of the
+    independence in the latent processes) can be specified quite generally by
+    only requiring the following covariances:
+     - Kuu: [L1, M, M]    (L can be 1, the conditional can broadcast),
```
Comment (vdutor) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
     - Kuu: [L/1, M, M]    (L can be 1, the conditional can broadcast),
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        self.inducing_variable = inducing_variable
+
+    def __len__(self):
+        return len(self.inducing_variable)
+
+
+class FallbackSeparateIndependentInducingVariables(MultioutputInducingVariables):
+    """
+    This class is designated to be used to
+     - provide a general a general interface for multioutput kernels
```
Comment (vdutor) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
     - provide a general interface for multioutput kernels
```

File: gpflow/covariances/mo_kuus.py (hunk 5 above)
Comment (markvdw) [reply gpflow/covariances/mo_kuus.py]: I'll leave this for later.

**Comment by markvdw** [comment]
I've changed a bunch of references to features in the tests now, but I'm stopping. There are loads, and it's a pain to do, and I don't want to break anything. I think it's ok for variable names in the tests to be outdated. We can update these if the tests ever break. Thoughts?

**Comment by markvdw** [comment]
I'm also using plural for referring to most inducing variables. I choose singular/plural based on whether it makes sense in a sentence.

File: gpflow/expectations/expectations.py [hunk 20]
```diff
@@ -23,8 +23,8 @@ def expectation(p, obj1, obj2=None, nghp=None):
     if one is available. If not, it falls back to quadrature.
 
     :type p: (mu, cov) tuple or a `ProbabilityDistribution` object
-    :type obj1: kernel, mean function, (kernel, features), or None
-    :type obj2: kernel, mean function, (kernel, features), or None
+    :type obj1: kernel, mean function, (kernel, inducing), or None
```
Comment (awav) [inline gpflow/expectations/expectations.py]:
```suggestion
    :type obj1: kernel, mean function, (kernel, inducing_variable), or None
```

File: gpflow/expectations/expectations.py [hunk 21]
```diff
@@ -23,8 +23,8 @@ def expectation(p, obj1, obj2=None, nghp=None):
...
+    :type obj2: kernel, mean function, (kernel, inducing), or None
```
Comment (awav) [inline gpflow/expectations/expectations.py]:
```suggestion
    :type obj2: kernel, mean function, (kernel, inducing_variable), or None
```

File: gpflow/expectations/expectations.py [hunk 22]
```diff
@@ -62,8 +62,8 @@ def quadrature_expectation(p, obj1, obj2=None, nghp=None):
     Uses Gauss-Hermite quadrature for approximate integration.
 
     :type p: (mu, cov) tuple or a `ProbabilityDistribution` object
-    :type obj1: kernel, mean function, (kernel, features), or None
-    :type obj2: kernel, mean function, (kernel, features), or None
+    :type obj1: kernel, mean function, (kernel, inducing), or None
```
Comment (awav) [inline gpflow/expectations/expectations.py]:
```suggestion
    :type obj1: kernel, mean function, (kernel, inducing_variable), or None
```

File: gpflow/expectations/expectations.py [hunk 23]
```diff
@@ -62,8 +62,8 @@ def quadrature_expectation(p, obj1, obj2=None, nghp=None):
...
+    :type obj2: kernel, mean function, (kernel, inducing), or None
```
Comment (awav) [inline gpflow/expectations/expectations.py]:
```suggestion
    :type obj2: kernel, mean function, (kernel, inducing_variable), or None
```

File: gpflow/conditionals/mo_conditionals.py [hunk 24]
```diff
@@ -177,10 +189,12 @@ def _conditional(Xnew, feature, kernel, f, *, full_cov=False, full_output_cov=Fa
...
 
-@conditional.register(object, (MixedKernelSharedMof, MixedKernelSeparateMof), SeparateMixedMok, object)
-def _conditional(Xnew, feature, kernel, f, *, full_cov=False, full_output_cov=False, q_sqrt=None, white=False):
+@conditional.register(object, (SharedIndependentInducingVariables, SeparateIndependentInducingVariables),
+                      LinearCoregionalization, object)
+def _conditional(Xnew, inducing_variable, kernel, f, *, full_cov=False, full_output_cov=False, q_sqrt=None,
+                 white=False):
     """Most efficient routine to project L independent latent gps through a mixing matrix W.
-    The mixing matrix is a member of the `SeparateMixedMok` and has shape [P, L].
+    The mixing matrix is a member of the `LinearCoregionalisation` and has shape [P, L].
```
Comment (awav) [inline gpflow/conditionals/mo_conditionals.py]:
```suggestion
    The mixing matrix is a member of the `LinearCoregionalization` and has shape [P, L].
```

File: gpflow/inducing_variables/mo_inducing_variables.py [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+
+class MultioutputInducingVariables(InducingVariables):
+    """
+    Multioutput Inducing Variables
+    Base class for methods which define a collection of inducing variables which
+    in some way can be grouped. The main example is where the inducing variables
+    consist of outputs of various independent GPs. This can be because our model
+    uses multiple independent GPs (SharedIndependent, SeparateIndependent) or
+    because it is constructed from independent GPs (eg IndependentLatent,
+    LinearCoregionalisation).
```
Comment (awav) [inline gpflow/inducing_variables/mo_inducing_variables.py]:
```suggestion
    LinearCoregionalization).
```

**Pull request merged**